### PR TITLE
Stronger https detection in Tools::getShopProtocol()

### LINKED ISSUE #28580: Harmonize http/https check
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

http/https check is not consistent

Sometimes `Configuration::get('PS_SSL_ENABLED')` is used or `Tools::usingSecureMode()`

`Tools::usingSecureMode()` tests if we are in https at server config level while `Configuration::get('PS_SSL_ENABLED')` tests if prestashop is https enabled in its config.

So I guess there should be a usecase where the server is in http and the config in https or the opposite ? Please help on this.

Sometimes `$_SERVER['HTTPS']` is checked and no other server env variables like HTTP_X_FORWARDED_PROTO

### Describe the solution you'd like

Harmonize http/https check in the code

### Alternatives you've considered

_No response_

### Additional context

_No response_

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by jf-viguier**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Stronger https detection in Tools::getShopProtocol(): Tools::usingSecureMode() check more server variables like HTTP_X_FORWARDED_PROTO.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/28580
| How to test?      | configure and httpd server with only HTTP_X_FORWARDED_PROTO var and see errors un https mode on.
| Possible impacts? | no
| Sponsor company   | Creabilis.com

### CODE REVIEW

**Comment by davd** [comment]
This change has a really small impact on the current core code base.

It seems that this function is sparsely used and the team is replacing calls to it. 

https://github.com/PrestaShop/PrestaShop/search?q=getShopProtocol

Some calls have been removed from the core and replaced by 

`Tools::getProtocol((bool) Configuration::get('PS_SSL_ENABLED'))`

see https://github.com/PrestaShop/PrestaShop/commit/311f3aabbd1c0328e96d4defc21b881f0c17c505

I think that marking this function as deprecated is a good idea as well.

File: classes/Tools.php
Comment (matks) [inline classes/Tools.php]: @jf-viguier Adding a return type is a BC break 😉 

If someone extends Tools class, and extend `getShopProtocol()`, then upgrades from 1.7.8 to 8.0.0, his class extend will crash because of your `: string` as the signature of the method will not match, so PHP will crash

**Review by matks — COMMENTED** [review]
Change is valid for me but please document the BC break or remove it 😉

File: classes/Tools.php
Comment (jf-viguier) [reply classes/Tools.php]: @matks BC break removed.

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
@PrestaShop/qa-team this is a refactoring PR, is it OK to test it by developers?

**Comment by HanaRebaiQA** [comment]
Hello @matks 

It's OK for us to be tested by developers and we should run the automatic tests on it.
Could you please check the fail on CI ?
Thank you!

**Comment by matks** [comment]
@jf-viguier I think you need to rebase, the CI has been fixed on `develop`

**Comment by jf-viguier** [comment]
@matks I've rebased

**Review by Progi1984 — DISMISSED** [review]

**Review by HanaRebaiQA — DISMISSED** [review]
Hello @jf-viguier 

I checked the automatic tests. It's OK :heavy_check_mark: 
@PrestaShop/prestashop-core-developers : Could you please test this PR and put the QA label to can merge it ?
And, thanks to do the second review.

Thank you!

**Review by FabienPapet — DISMISSED** [review]

**Comment by FabienPapet** [comment]
Friendly reminder @PrestaShop/qa-team

**Comment by HanaRebaiQA** [comment]
Hello @FabienPapet 

I have launched the automatic tests on this PR and it's QA :heavy_check_mark: 
And now, it's up to developers to test it and approve it.

Thank You!

**Comment by jf-viguier** [comment]
@FabienPapet ?

**Comment by HanaRebaiQA** [comment]
Friendly reminder @PrestaShop/prestashop-core-developers

**Comment by jf-viguier** [comment]
rebase done @FabienPapet

**Comment by Progi1984** [comment]
Ping @PrestaShop/prestashop-maintainers The rebase has been done by @jf-viguier

**Comment by FabienPapet** [comment]
Good for me

**Review by FabienPapet — APPROVED** [review]

**Review by eternoendless — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
@PrestaShop/qa-team 

Automated tests should be enough if we can run them using `https`

**Comment by kpodemski** [comment]
Ok, I think this one, eventually, should be tested by the dev. Waiting for confirmation from QA.

**Comment by AureRita** [comment]
Hello @kpodemski 

Yes, I think this PR should be tested by the dev 

Thank you

**Comment by jf-viguier** [comment]
Any news on this PR ?

**Comment by atomiix** [comment]
Did some tests, okay for me, QA ✔️.
However, it brought some security concerns to my attention as without a proper apache config it's actually possible to bypass the SSL redirection using `X-Forwarded-Proto` header.

**Comment by atomiix** [comment]
Thanks @jf-viguier

**Pull request merged**
